### PR TITLE
CLI scripts

### PR DESCRIPTION
**Pull request by chaithyagr**
This is an ongoing PR for creating CLI for pysap-mri reconstruct to streamline and make the work reproducibile.

### CODE REVIEW

File: src/mri/cli/dc_adjoint.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    traj_file : str
+        Path to the trajectory file.
+    obs_reader : callable
+        A function that reads the observed data file and returns
+        the raw data and data header.
+    traj_reader : callable
+        A function that reads the trajectory file and returns the trajectory
+        data and parameters.
+    fourier : callable
+        A function that performs Fourier transform operations.
```
Comment (paquiteau) [inline src/mri/cli/dc_adjoint.py]:
```suggestion
fourier: Callable
   A Callable returning a Fourier Operator
``` 
Also described what input argument it should have.

File: src/mri/cli/dc_adjoint.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    -------
+    None
+        The reconstructed image is saved as 'dc_adjoint.pkl' file.
+    """
+    raw_data, data_header = obs_reader(obs_file)
+    if "trajectory_name" in data_header.keys():
+        if data_header["trajectory_name"] != os.path.basename(traj_file):
+            log.warn("Trajectory file does not match the trajectory in the data file")
+    else:
+        log.warn("Trajectory name not found in data header, validation step not done")
```
Comment (paquiteau) [inline src/mri/cli/dc_adjoint.py]:
```suggestion
    try:
        if data_header["trajectory_name"] != os.path.basename(traj_file):
            log.warn("Trajectory file does not match the trajectory in the data file")
    except KeyError:
        log.warn("Trajectory name not found in data header, Skipped Validation")
```
"Easier to ask for forgiveness than permission"

File: src/mri/cli/dc_adjoint.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    shots, traj_params = traj_reader(
+        traj_file, dwell_time=DEFAULT_RASTER_TIME / data_header["oversampling_factor"]
```
Comment (paquiteau) [inline src/mri/cli/dc_adjoint.py]: you may want to have the raster time being a parameter for future flexibility.

File: src/mri/cli/dc_adjoint.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    zen_exclude=["n_coils"],
+)
+fourier_store = store(group="fourier")
+fourier_store(fourier_op_config, name="cpu")
+fourier_store(
+    fourier_op_config, implementation="gpuNUFFT", density_comp="pipe", name="gpu"
+)
+
+
+def recon(obs_file: str, traj_file: str, obs_reader, traj_reader, fourier):
```
Comment (paquiteau) [inline src/mri/cli/dc_adjoint.py]: Type hints for hydra zen to be happy ?

File: src/mri/cli/dc_adjoint.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    )
+    log.debug(f"Normalized shifts: {normalized_shifts}")
+    kspace_data = add_phase_to_kspace_with_shifts(
+        np.squeeze(raw_data), kspace_loc, normalized_shifts
+    )
+    fourier_op = fourier(
+        kspace_loc, traj_params["img_size"], n_coils=data_header["n_coils"]
+    )
+    per_ch_image = fourier_op.adj_op(kspace_data)
+    combined_image = np.linalg.norm(per_ch_image, axis=-1)
```
Comment (paquiteau) [inline src/mri/cli/dc_adjoint.py]: what if I provided the smaps in the file or separately ?

File: src/mri/cli/dc_adjoint.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    pkl.dump(combined_image, open("dc_adjoint.pkl", "wb"))
```
Comment (paquiteau) [inline src/mri/cli/dc_adjoint.py]: I would rather have the output format left to the user (.npy, .nii)m, as well as the filename. Once reconstructed, the user will probably want to view the data in their favorite viewer.

**Review by paquiteau — COMMENTED** [review]
Nice start, later examples on how to use the CLI would be nice.

File: src/mri/cli/dc_adjoint.py (hunk 4 above)
Comment (chaithyagr) [reply src/mri/cli/dc_adjoint.py]: That is handeled above. The readers are partially configured objects based on inputs.

**Review by paquiteau — APPROVED** [review]
LGTM, we will see how usage would dictate new changes ..

File: src/mri/cli/dc_adjoint.py (hunk 3 above)
Comment (chaithyagr) [reply src/mri/cli/dc_adjoint.py]: It already is, will update the code to bring that config here. Good catch

File: src/mri/cli/dc_adjoint.py (hunk 5 above)
Comment (chaithyagr) [reply src/mri/cli/dc_adjoint.py]: Ya, thats not yet supported :P

File: pyproject.toml [hunk 7]
```diff
@@ -27,6 +27,11 @@ requires-python = ">=3.10"
 [project.optional-dependencies]
 test=['pytest>=5.0.1', 'pytest-cov>=2.7.1', 'pytest-pep8', 'pytest-runner', "pytest-xdist", "pytest-sugar"]
 gpu=["cupy", "cufinufft", "gpunufft"]
+scripts=["hydra-zen", "pymrt", "nibabel", "hydra-callbacks"]
+
+[project.scripts]
+recon = "mri.cli.reconstruct:run_recon"
+dc_adjoint = "mri.cli.reconstruct:run_adjoint"
```
Comment (chaithyagr) [inline pyproject.toml:34]: These are the CLI scripts.

**Pull request merged**